src/game/behaviors/fish.inc.c

```diff
@@ -1,163 +1,299 @@
     }
-    o->oFishGroupUnkF8 = gMarioObject->oPosY + o->oFishGroupUnkFC;
+    
+    // Rotate fish towards mario at a rate of 5.625 degrees (0x400)
+    o->oFishPosY = gMarioObject->oPosY + o->oFishRandomOffset;
     obj_rotate_yaw_toward(o->oAngleToMario, 0x400);
-    if (o->oPosY < o->oFishGroupUnkF4 - 50.0f) {
-        if (sp1C < 0.0f)
-            sp1C = 0.0f - sp1C;
-        if (sp1C < 500.0f)
-            func_802BF59C(2);
-        else
-            func_802BF59C(4);
+    
+    // If fish groups are too close, call fish_regroup()
+    if (o->oPosY < o->oFishWaterLevel - 50.0f) {
+        if (fishY < 0.0f) {
+            fishY = 0.0f - fishY;
+        }
+        if (fishY < 500.0f) {
+            fish_regroup(2);
+        } else {
+            fish_regroup(4);
+        }
     } else {
-        o->oPosY = o->oFishGroupUnkF4 - 50.0f;
-        if (sp1C > 300.0f)
+        o->oPosY = o->oFishWaterLevel - 50.0f;
+        if (fishY > 300.0f) {
             o->oPosY = o->oPosY - 1.0f;
+        }
+    }
+    
+    /**
+     * Delete current fish but create a new one if distance to mario is
+     * smaller than his distance to oFishRandomDistance + 150.0f.
+     */
+    if (o->oDistanceToMario < o->oFishRandomDistance + 150.0f) {
+        o->oAction = FISH_RESPAWN;
     }
-    if (o->oDistanceToMario < o->oFishGroupUnk104 + 150.0f)
-        o->oAction = 2;
 }
 
-void ActionFishGroup2(void) {
-    f32 sp1C = o->oPosY - gMarioObject->oPosY;
-    s32 sp18;
-    o->oFishGroupUnkF8 = gMarioObject->oPosY + o->oFishGroupUnkFC;
+/**
+ * Moves fish in relation to distance from each other and Mario.
+ */
+void fish_group_act_move(void) {
+    f32 fishY = o->oPosY - gMarioObject->oPosY;
+    // Marked unused, but has arithmetic performed on it in a useless manner.
+    UNUSED s32 distance;
+    o->oFishPosY = gMarioObject->oPosY + o->oFishRandomOffset;
+    /**
+     * Set fish variables to random floats when timer reaches zero and plays sound effect.
+     * This allows fish to move naturally.
+     */
     if (o->oTimer == 0) {
-        o->oFishGroupUnk110 = RandomFloat() * 300.0f;
-        o->oFishGroupUnk100 = RandomFloat() * 1024.0f + 1024.0f;
-        o->oFishGroupUnk108 = RandomFloat() * 4.0f + 8.0f + 5.0f;
-        if (o->oDistanceToMario < 600.0f)
-            sp18 = 1;
-        else
-            sp18 = (s32)(1.0 / (o->oDistanceToMario / 600.0));
-        sp18 *= 127;
+        o->oFishRandomDistance2 = RandomFloat() * 300.0f;
+        o->oFishRandomSpeed = RandomFloat() * 1024.0f + 1024.0f;
+        o->oFishRandomVel = RandomFloat() * 4.0f + 8.0f + 5.0f;
+        if (o->oDistanceToMario < 600.0f) {
+            distance = 1;
+        } else {
+            distance = (s32)(1.0 / (o->oDistanceToMario / 600.0));
+        }
+        distance *= 127;
         PlaySound2(SOUND_GENERAL_MOVING_WATER);
     }
-    if (o->oTimer < LEVEL_SA)
+    // Enable fish animation. Parameters differ if oTimer < SA.
+    if (o->oTimer < 20) {
```

Comment:
Oh is this why you changed LEVEL_SA earlier? You were right to change it here, but before, it was a proper reference to the level

Reply:
Did I misunderstand that all "LEVEL_SA" should just be 20?

Reply:
yes, he was only talking about that one, since it's obviously not a level reference.

Just think about it: which of these makes sense?
`myObject.position == POSITION_VECTOR_1`
`myObject.position == 20`
`myObject.speed == POSITION_VECTOR_1`
`myObject.speed == 20`
replace position with level and speed with timer and it's the same situation